packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectQuery.kt

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2021 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.realm.internal
+
+import io.realm.RealmObject
+import io.realm.RealmResults
+import io.realm.internal.interop.NativePointer
+import io.realm.internal.interop.RealmCoreException
+import io.realm.internal.interop.RealmCoreIndexOutOfBoundsException
+import io.realm.internal.interop.RealmCoreInvalidQueryException
+import io.realm.internal.interop.RealmCoreInvalidQueryStringException
+import io.realm.internal.interop.RealmInterop
+import io.realm.query.RealmQuery
+import io.realm.query.RealmScalarQuery
+import io.realm.query.RealmSingleQuery
+import io.realm.query.Sort
+import kotlinx.coroutines.flow.Flow
+import kotlin.reflect.KClass
+
+@Suppress("SpreadOperator")
+internal class RealmObjectQuery<E : RealmObject> constructor(
+    private val realmReference: RealmReference,
+    private val clazz: KClass<E>,
+    private val mediator: Mediator,
+    composedQueryPointer: NativePointer? = null,
+    private val filter: String,
+    private vararg val args: Any?
+) : RealmQuery<E>, Thawable<RealmResultsImpl<E>> {
+
+    private val queryPointer: NativePointer = when {
+        composedQueryPointer != null -> composedQueryPointer
+        else -> parseQuery()
+    }
+
+    private val resultsPointer: NativePointer by lazy {
```

Comment:
This doesn't seem properly frozen. I get that it will work in a lot of scenarios, but constructing a query on one thread and passing it to another to be processed (ex. with`find()`) would cause mutability issues on native.  

So I guess there isn't a test for that :)

Reply:
There is: https://github.com/realm/realm-kotlin/pull/603/files#diff-4e7ed4ec30aad17f1a9fb5c70a9035859e8eecf7b61728001177f403d5f7a7edR762 - but it's commented out because of the PR split (`count` is not implemented yet, so I left it there for later). I will uncomment it and remove the references to `count`, but as far as I can see it runs just fine on native.

Please check whether the logic makes sense or not, but I made sure to create the query on a thread and call `find()` on a different one and there aren't any crashes due to mutability.

Reply:
Seems like the `lazy` is properly _freeze aware_ and implemented on top of atomic references 🎉